src/helper/engine.cpp

```diff
@@ -58,12 +58,24 @@ void engine::update(engine_ruleset &ruleset,
     std::atomic_store(&common_, new_common);
 }
 
-std::optional<engine::result> engine::context::publish(parameter &&param)
+std::optional<engine::result> engine::context::publish(
+    parameter &&param, request_stage stage)
 {
     // Once the parameter reaches this function, it is guaranteed to be
     // owned by the engine.
     prev_published_params_.push_back(std::move(param));
 
+    std::optional<sampler::scope> opt_scope;
+    if (schema_extraction_enabled_ && stage == request_stage::shutdown) {
+        opt_scope = schema_sampler_.get();
+        if (opt_scope.has_value()) {
+            parameter &root = prev_published_params_.back();
+            parameter context_processor = parameter::map();
+            context_processor.add("extract-schema", parameter::boolean(true));
+            root.add("waf.context.processor", std::move(context_processor));
+        }
+    }
```

Comment:
On balance, I'd move this logic to the caller. You're adding new responsibilities the engine::context besides merely distributing waf data; now you're conditionally modifying this data as well. Also now you don't have all the information to know if you should do it (if it's the right moment in the request), so you then modify the interface of the class to provide the stage as well, and the actual action being performed is hidden behind the `if`. OTOH the caller's responsibility is already to provide waf data, and it knows already what stage it's processing.

In general, if you have a function `do_x()`, and now you pass it a new parameter to sometimes do something nonobvious you should think twice.

Reply:
I like your suggestion. I am not against it. However the reason why this code landed here was because of the sampler. The caller of this function is the client and the sampler does not belong it there. I don't mind to keep the sampler here and ask for the client if the request is selected to the engine, however that mean that client and engine would be aware of schema extraction. I'm ok both approaches but I prefer we agree in one first and then I'll do the changes